### PR TITLE
fix(provider): stop the npm build turning new.target into import.meta

### PR DESCRIPTION
**Pull request by kojiwakayama**
## The log line

Found by running the published `veryfront@0.1.1232` artifact from npm, not by reading code:

```
err=undefined: openai request failed: Provider request failed with status 400
```

The literal `undefined` sits where the error's `name` belongs.

## Mechanism

DNT rewrites `import.meta` into its ESM ponyfill by visiting **meta-property** AST nodes. `new.target` is also a meta-property, and the transform does not distinguish the two.

`src/provider/runtime-loader/provider-http.ts:66` reads `this.name = new.target.name`. In the published tarball, `esm/src/provider/runtime-loader/provider-http.js:44` reads:

```js
this.name = globalThis[Symbol.for("import-meta-ponyfill-esmodule")](import.meta).name;
```

The ponyfill returns an `ImportMeta` — `url`, `resolve`, never `name` — so `this.name` is `undefined` on every `ProviderError` the shipped package throws.

Reproduced against the published package:

```
typeof e.name : undefined
e.name        : undefined

ACTUAL LOG LINE:
  err=undefined: openai request failed: Provider request failed with status 400
```

## Same root cause, second symptom

The four filesystem adapters guard with `new.target === SomeClass`. In the published package that becomes `ponyfill(import.meta) === SomeClass`, unconditionally false — so `markNativeFileSystemAdapter` never ran in any shipped build:

```js
// esm/src/platform/adapters/runtime/node/filesystem-adapter.js:8
if (globalThis[Symbol.for("import-meta-ponyfill-esmodule")](import.meta) === NodeFileSystemAdapter) {
```

Third failure mode: on an import path that never loads `_dnt.polyfills.js`, the rewritten expression throws `TypeError: globalThis[Symbol.for(...)] is not a function` instead of returning a wrong value.

## Fix

`this.constructor` at all five sites. Not a meta-property, so DNT leaves it alone. Differs from `new.target` only under `Reflect.construct(Base, args, Other)`, which this repo does not do.

## Test

The damage is invisible to the Deno suite — the sources are correct, the *emitted package* is not — so the guard is a source-level audit, `scripts/build/dnt-meta-property-safety.ts`, matching on the `MetaProperty` AST node (a mention in a comment or string is not reported).

Red before the fix, with exactly the five real sites:

```
finds no new.target anywhere in the shipped sources ... FAILED (2s)

error: AssertionError: Values are not equal: DNT rewrites new.target into the
import.meta ponyfill, so these are silently broken in the published npm package.

-   [
-     "src/platform/adapters/runtime/deno/filesystem-adapter.ts:159",
-     "src/platform/adapters/runtime/shared/node-filesystem-adapter.ts:435",
-     "src/platform/adapters/runtime/bun/filesystem-adapter.ts:24",
-     "src/platform/adapters/runtime/node/filesystem-adapter.ts:12",
-     "src/provider/runtime-loader/provider-http.ts:66",
-   ]
+   []
```

Green after: `ok | 1 passed (7 steps) | 0 failed (1s)`.

Also green: `src/provider/runtime-loader/provider-http.test.ts` (50 steps), `src/platform/adapters/runtime/` (386 steps), plus the full pre-push suite.

## Not fixed here

This restores *which* error it was. It does not add the provider's own 400 body to the message — that is deliberate (`does not surface provider error body contents`, provider-http.test.ts:260) and is handled in the attachment PR that follows.

Do not merge or queue — review only.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved runtime adapter detection across supported environments, preserving correct native-adapter behavior.
  - Ensured subclasses and altered prototypes are not incorrectly identified as native adapters.
  - Preserved accurate provider error names in generated packages.
  - Improved compatibility by eliminating unsafe metadata usage from shipped sources.

- **Documentation**
  - Corrected source references in the provider API documentation.

- **Tests**
  - Added automated checks for metadata safety, invalid source handling, shipped-code compatibility, and adapter detection edge cases.

### CODE REVIEW

File: scripts/build/dnt-meta-property-safety.ts [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+      }
+    }
+  };
+
+  visit(ast.program as unknown as Node);
+  return uses.sort((a, b) => a.line - b.line);
+}
+
+/** Source roots copied into the npm package by the DNT build. */
+export const SHIPPED_SOURCE_ROOTS = ["src", "cli", "react"] as const;
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply scripts/build/dnt-meta-property-safety.ts]: Fixed in 29a3be2 — you were right, `templates/` and `extensions/` both ship through DNT and both were outside the scan.

The roots are now derived instead of listed (`shippedSourceRoots`, scripts/build/dnt-meta-property-safety.ts):

- root package: `build-npm-dnt.ts:61` takes `entryPoints` straight from `deno.json`'s export map, so every export's top-level directory is scanned. `"./scaffold": "./templates/scaffold.ts"` (deno.json:177) is what pulled `templates/` in.
- extension packages: `build-npm-extension-packages.ts` runs one DNT build per first-party `./extensions/*` workspace member (`firstPartyExtensionManifestPaths`, npm-extension-package-metadata.ts:122), so each of those 28 directories is scanned.
- `react/` stays explicit: those are the shim modules the root build maps onto the bare `react`/`react-dom` specifiers.

Derived roots now: `cli`, `react`, `src`, `templates`, and `extensions/ext-*` (28). A new test asserts that every value in the export map and every `./extensions/*` workspace member falls inside an audited root, so adding an export in a new directory fails the audit rather than slipping past it.

File: deno.json [hunk 2]
```diff
@@ -543,7 +543,7 @@
...
...ts/build/npm-extension-package-metadata.test.ts scripts/build/npm-package-metadata.test.ts scripts/build/npm-react-shims.test.ts scripts/build/npm-runtime-helper-contract.test.ts scripts/build/prepare-framework-sources.test.ts scripts/docs/docs-coverage.test.ts scripts/docs/generate-api-reference.test.ts scripts/docs/guide-validation.test.ts scripts/lint/audit-chat-composability.test.ts scripts/lint/audit-rfc-status.test.ts scripts/lint/audit-core-deps.test.ts scripts/lint/audit-cwd-relative-test-reads.test.ts scripts/lint/audit-cross-runtime-jsr.test.ts scripts/lint/audit-dependency-boundaries.test.ts scripts/lint/audit-extension-capabilities.test.ts scripts/lint/audit-extension-contracts.test.ts scripts/lint/audit-deps.test.ts scripts/lint/check-module-boundaries.test.ts scripts/lint/lint-config.test.ts scripts/lint/ban-test-only.test.ts scripts/lint/check-sanitizer-baseline.test.ts scripts/lint/check-skipped-tests-baseline.test.ts scripts/lint/check-test-typecheck-baseline.test.ts scripts/lint/check-coverage.test.ts scripts/security/audit-npm.test.ts scripts/security/submit-dependency-snapshot.test.ts scripts/test/template-runtime-e2e.test.ts && deno task test:tool-search-live",
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply deno.json:547]: Fixed in 29a3be2. Confirmed your reading of the workflow: `deno task test:scripts` appears in no job in `.github/workflows/cicd.yml`, and `build:npm` is called directly at cicd.yml:724 and cicd.yml:890.

The audit now has its own task and runs in both places:

- `"lint:dnt-meta-properties": "deno run --config=scripts/test.deno.json --frozen --allow-read scripts/build/dnt-meta-property-safety.ts"`
- `build:npm` runs it immediately before `build-npm-dnt.ts`, so a publish cannot emit a corrupted package.
- `lint:ci` runs it too, so a PR fails on the audit rather than only at release time (cicd.yml:50 dispatches `lint` to `deno task lint:ci`).

It is read-only and takes ~2.6s over the widened root set. The test stays in `test:scripts` as separate validation of the audit itself.

File: deno.json [hunk 3]
```diff
@@ -494,10 +494,10 @@
...
-    "fmt:check": "deno fmt --check src/ cli/ react/ templates/ && deno fmt --check --config=scripts/test.deno.json scripts/test/ scripts/build/dnt-polyfill.ts scripts/build/dnt-polyfill.test.ts scripts/build/prepare-framework-sources.test.ts && deno fmt --check --config=scripts/codemods/deno.json scripts/codemods/",
+    "fmt": "deno fmt src/ cli/ react/ templates/ && deno fmt --config=scripts/test.deno.json scripts/test/ scripts/build/dnt-meta-property-safety.ts scripts/build/dnt-meta-property-safety.test.ts scripts/build/dnt-polyfill.ts scripts/build/dnt-polyfill.test.ts scripts/build/prepare-framework-sources.test.ts && deno fmt --config=scripts/codemods/deno.json scripts/codemods/",
+    "fmt:check": "deno fmt --check src/ cli/ react/ templates/ && deno fmt --check --config=scripts/test.deno.json scripts/test/ scripts/build/dnt-meta-property-safety.ts scripts/build/dnt-meta-property-safety.test.ts scripts/build/dnt-polyfill.ts scripts/build/dnt-polyfill.test.ts scripts/build/prepare-framework-sources.test.ts && deno fmt --check --config=scripts/codemods/deno.json scripts/codemods/",
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply deno.json:500]: Fixed in 29a3be2. Confirmed your reading of the workflow: `deno task test:scripts` appears in no job in `.github/workflows/cicd.yml`, and `build:npm` is called directly at cicd.yml:724 and cicd.yml:890.

The audit now has its own task and runs in both places:

- `"lint:dnt-meta-properties": "deno run --config=scripts/test.deno.json --frozen --allow-read scripts/build/dnt-meta-property-safety.ts"`
- `build:npm` runs it immediately before `build-npm-dnt.ts`, so a publish cannot emit a corrupted package.
- `lint:ci` runs it too, so a PR fails on the audit rather than only at release time (cicd.yml:50 dispatches `lint` to `deno task lint:ci`).

It is read-only and takes ~2.6s over the widened root set. The test stays in `test:scripts` as separate validation of the audit itself.

File: scripts/build/dnt-meta-property-safety.ts [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+  for await (const entry of entries) {
+    const path = `${root}/${entry.name}`;
+    if (entry.isDirectory) {
+      if (entry.name === "node_modules" || entry.name.startsWith(".")) continue;
+      files.push(...await collectShippedSources(path));
+    } else if (SOURCE_EXTENSIONS.some((ext) => entry.name.endsWith(ext))) {
+      files.push(path);
+    }
+  }
+  return files;
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply scripts/build/dnt-meta-property-safety.ts:207]: Fixed in 29a3be2 — confirmed. `Deno.readDir()` builds the async iterable without touching the filesystem, so the old `try` around the call caught nothing and a missing root rejected out of the `for await` and out of `auditRepoMetaProperties`.

`collectShippedSources` now wraps the iteration, ignores only `Deno.errors.NotFound`, and rethrows everything else, so an unreadable root fails the audit instead of silently shrinking the scanned set. Three tests cover it:

- `collectShippedSources` on an absent directory returns `[]`;
- `collectShippedSources` on a path that is a file rejects (`NotADirectory` — a non-`NotFound` failure that must propagate);
- `auditRepoMetaProperties` against a temporary repo root whose `deno.json` names `./src/index.ts` but which has no `src/` returns no uses and no parse failures. That last one fails on the pre-fix code with an uncaught `NotFound`.

File: src/platform/adapters/runtime/shared/node-filesystem-adapter.ts [hunk 5]
```diff
@@ -432,7 +432,7 @@ export class NodeCompatibleFileSystemAdapter implements FileSystemAdapter {
         enumerable: true,
       });
     }
-    if (new.target === NodeCompatibleFileSystemAdapter) {
+    if (this.constructor === NodeCompatibleFileSystemAdapter) {
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/platform/adapters/runtime/shared/node-filesystem-adapter.ts]: The forgeability half is right and is fixed in 29a3be2. The suggested spelling — `new.target === <built-in class>` — is not, and taking it would reintroduce exactly the bug this PR exists to fix. Details, then what landed.

**Why `new.target` cannot come back.** DNT rewrites *meta-properties* into its `import.meta` ponyfill, and it does not distinguish `import.meta` from `new.target`. In the published veryfront@0.1.1232 tarball, `esm/src/provider/runtime-loader/provider-http.js:44` reads:

```js
this.name = globalThis[Symbol.for("import-meta-ponyfill-esmodule")](import.meta).name;
```

from source `this.name = new.target.name`. Applied to these four guards, `new.target === NodeCompatibleFileSystemAdapter` becomes `ponyfill(import.meta) === NodeCompatibleFileSystemAdapter`, which is unconditionally `false`: in the published npm build **no adapter is ever registered as native at all**. Every local test passes while the shipped artifact is inert — which is why the guard needs a spelling that survives the transform.

**What landed** — your second suggestion, the captured prototype-identity check. New `isDirectConstruction` in `src/platform/adapters/native-file-system-provenance.ts`, used at all four sites (shared, Node, Bun, Deno):

```ts
return ReflectApply(ObjectGetPrototypeOf, undefined, [instance]) === constructorRef.prototype;
```

A class's `prototype` is a non-writable, non-configurable own property, and `[[Construct]]` takes the new object's prototype from `new.target.prototype`, so `new Subclass()` always arrives carrying `Subclass.prototype`. Deleting or overwriting `prototype.constructor` changes nothing. It is an ordinary member expression, so DNT leaves it alone.

**Regressions**, one per adapter (shared/node/bun/deno `filesystem-adapter.test.ts`): a subclass with `Reflect.deleteProperty(Sub.prototype, "constructor")` and a subclass whose `prototype.constructor` is redefined to the base class — both must not be classified as native. Verified they fail against the `this.constructor` version and pass against `isDirectConstruction`.

**On the `Reflect.construct` equivalence**, checked against this repo rather than asserted: `Reflect.construct` appears at `src/platform/compat/std/testing/time.ts:231` (a `Date` proxy), `src/extensions/runtime-validation.ts:30` (`Reflect.construct(Object, [], value)`), and three `ArrayBuffer` calls in tests — none constructs a filesystem adapter or a `ProviderError`, and none passes a third argument that would separate `new.target` from the prototype the instance receives. The only in-repo subclasses of these five classes are `BunFileSystemAdapter`/`NodeFileSystemAdapter` (which register themselves) and the four `ProviderError` subclasses at provider-http.ts:81-90, which want their own name and get it from `this.constructor.name`.

`ProviderError`'s `this.name = this.constructor.name` is left as is: reading a name is not an identity test, and a subclass that hides its own `constructor` would merely report the base name.

**Comment by kojiwakayama** [comment]
### CI status

Both failing checks are green on `29a3be2`; 28 pass / 7 skipped / 0 fail.

**`ci (lint)` — real.** `docs/api-reference is stale: outdated: veryfront/provider.md`. The explanatory comment added to `provider-http.ts` shifted the line pins the generated reference records (`ProviderError` L47 unchanged, `buildProviderError` L163→167, `requestJson` L662→666, and the four subclasses). Regenerated with `deno task docs` on the pinned Deno 2.7.7 and committed; `deno task docs:api-reference:check` now reports the reference current.

**`tests (bun)` — real failure text, but not from this branch: a pre-existing nondeterministic test.**

```
(fail) child-run-execution-support > throwIfChildRunAborted > throws the signal Error reason when present
error: Expected error message to include "custom reason", got "The operation was aborted"
```

`src/agent/child-run/execution-support.test.ts` is not reachable from anything this PR touches — the diff is the five construction sites, `deno.json` task strings, and two new files under `scripts/`. `"The operation was aborted"` is the `DOMException` fallback in `createAbortError` (src/utils/abort.ts:19), taken when `isErrorAcrossRealms(signal.reason)` is false; the very next test in the same file asserts strict identity on an Error reason and passed in the same run, which is inconsistent with a deterministic fault.

Evidence gathered before re-running rather than after:
- the same file passes locally under Bun 1.3.6 through the real runner (`node ./tests/bun/run-tests.mjs`) after a full `deno task build:npm`, and standalone under the CI env (`DENO_TESTING=1 NODE_ENV=production …`);
- main passed `tests (bun)` at this PR's merge base 688cb7f9 with the identical file count (1312 vs our 1311+1);
- re-running the job unchanged on the same commit passed (job 94370388221).

Not fixed here — it is a latent flake in `throwIfChildRunAborted` under Bun on Linux that belongs in its own change, and touching it from this PR would hide it.

### Declined (CodeRabbit nitpick, `filesystem-adapter.ts` imports)

The suggestion to rewrite `../../native-file-system-provenance.ts` as `#veryfront/platform/adapters/native-file-system-provenance.ts` does not apply: `runtime/bun`, `runtime/deno`, `runtime/node`, `runtime/shared` and the provenance module all live inside `src/platform/adapters/`, so the import crosses no module boundary. The quoted guideline is "use `#veryfront/*` aliases only when an import crosses a module boundary", and all four files already used the relative form before this PR — the diff only widened the named-import list.
